java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java

```diff
@@ -348,17 +355,41 @@ Optional<AccessToken> getSharedAccessTokenIfSupported() {
         }
 
         @VisibleForTesting
-        void storeSharedAccessTokenIfSupported(@NonNull AccessToken accessToken) {
+        AccessToken storeSharedAccessTokenIfSupported(@NonNull AccessToken newToken) {
+            AccessToken result = newToken;
             if (payloadBuilder.useSharedToken()) {
-                try {
-                    config.putConfigProperty(ConfigProperty.ACCESS_TOKEN,
-                        objectMapper.writerFor(AccessTokenDto.class)
-                            .writeValueAsString(AccessTokenDto.toAccessTokenDto(accessToken)));
-                    log.log(Level.INFO, "New token stored in config");
-                } catch (JsonProcessingException e) {
-                    log.log(Level.SEVERE, "Could not serialize token into JSON", e);
+                // at least synchronize in the same JVM
+                synchronized ($refreshLock) {
+                    try {
+                        // wait slightly before checking for most recent token in config
+                        Uninterruptibles.sleepUninterruptibly(RandomUtils.nextLong(50, 200), TimeUnit.MILLISECONDS);
```

Comment:
again, already a `RandomNumberGenerator` being injected above; use that, can fake it to 1 for tests or something